R/init.R

```diff
@@ -122,52 +121,59 @@ init <- function(data,
       )
     )
   }
-
+  if (is.list(data) && !inherits(data, "teal_data_module")) {
+    checkmate::assert_list(data, names = "named")
+    data <- do.call(teal.data::teal_data, data)
+  }
   checkmate::assert_multi_class(data, c("teal_data", "teal_data_module"))
-  checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules"))
+
+  ## `modules`
+  if (inherits(modules, "teal_module")) {
+    modules <- list(modules)
+  }
+  if (inherits(modules, "list")) {
+    modules <- do.call(teal::modules, modules)
+  }
+  checkmate::assert_class(modules, "teal_modules")
+
+  ## `filter`
   checkmate::assert(
     checkmate::check_class(filter, "teal_slices"),
     checkmate::check_list(filter, names = "named")
   )
+
+  ## all other arguments
   checkmate::assert_multi_class(title, c("shiny.tag", "character"))
   checkmate::assert_multi_class(header, c("shiny.tag", "character"))
   checkmate::assert_multi_class(footer, c("shiny.tag", "character"))
   checkmate::assert_character(id, max.len = 1, any.missing = FALSE)
 
+  # log
   teal.logger::log_system_info()
 
-  if (is.character(title)) {
-    title <- build_app_title(title)
-  } else {
-    validate_app_title_tag(title)
-  }
-
-  if (inherits(modules, "teal_module")) {
-    modules <- list(modules)
-  }
-  if (inherits(modules, "list")) {
-    modules <- do.call(teal::modules, modules)
-  }
-
+  # argument mutation
+  ## `modules` - landing module
   landing <- extract_module(modules, "teal_module_landing")
-  if (length(landing) > 1L) stop("Only one `landing_popup_module` can be used.")
+  landing_module <- NULL
+  if (length(landing) > 1L) {
+    stop("Only one `landing_popup_module` can be used.")
+  } else if (length(landing) == 1) {
+    landing_module <- landing[[1L]]
+  }
   modules <- drop_module(modules, "teal_module_landing")
+  rm(landing)
 
+  ## `filter` - app_id attribute
   # Calculate app id that will be used to stamp filter state snapshots.
   # App id is a hash of the app's data and modules.
   # See "transferring snapshots" section in ?snapshot.
```

Comment:
Since this is delegated to a function, these comments will be more appropriate in that function's documentation.
```suggestion
```